### PR TITLE
Revamp doc intro 

### PR DESCRIPTION
**Pull request by harryskim**
Edited introduction for nGraph documentation with updated content and diagrams.

### CODE REVIEW

File: doc/sphinx/source/project/introduction.rst [hunk 1]
```diff
@@ -4,176 +4,86 @@
...
-an enormously complex and difficult-to-scale pipeline (:ref:`Figure C <figure-C>`) 
-from "training with your favorite framework using GPUs" (:ref:`Figure D <figure-D>`), 
-to deploying that (now) pre-trained model in a datacenter or production 
-environment, where infrastructure owners or software developers renting anything 
-in a datacenter ought to be mutually concerned with **efficiency per-watt**, to 
-keep costs in check.
-
-So what exactly are the motivations behind the nGraph Compiler stack? 
+Accelerating deep learning performance is critical for further development of Artificial Intelligence (AI). Deep learning models & corresponding data set sizes are increasing rapidly, and deployment of scalable AI solution is challenging. Currently, there are two general approaches to advancing deep learning performance. First approach is to design hardware solutions dedicated to deep learning computation. Exisiting large silicon companies including Intel and deep learning hardware startups are actively developing ASICs (Application Specific Integrated Circuit) for speeding up deep learning performance for training & inference.
```
Comment (indie) [inline doc/sphinx/source/project/introduction.rst]: Adding @aprocter to help review.  On first glance, the rewrite needs some work; let's make sure the engineering details are accurate.  The first sentence, for example, is not really accurate; today nGraph has only experimental support for "accelerating" distributed training workloads, so opening with a sentence that makes it sound like that's what we're doing here could be confusing.

Comment (indie) [inline doc/sphinx/source/project/introduction.rst]: Replace all instances of  `&` with "and"

Comment (indie) [inline doc/sphinx/source/project/introduction.rst]: The convention in our and other ML research is to call "data sets" "datasets"

Comment (indie) [inline doc/sphinx/source/project/introduction.rst]: performance. The first approach is....

Comment (indie) [inline doc/sphinx/source/project/introduction.rst]: > Exisiting large silicon companies including Intel and deep learning hardware startups are actively developing ASICs (Application Specific Integrated Circuit) for speeding up deep learning performance for training & inference.   

While this is good info to have in the docs, I'm not sure we want to lead with that info in the documentation's introduction.  Info like this might be more useful in a blog.

File: doc/sphinx/source/project/introduction.rst [hunk 2]
```diff
@@ -4,176 +4,86 @@
...
+Accelerating deep learning performance is critical for further development of Artificial Intelligence (AI). Deep learning models & corresponding data set sizes are increasing rapidly, and deployment of scalable AI solution is challenging. Currently, there are two general approaches to advancing deep learning performance. First approach is to design hardware solutions dedicated to deep learning computation. Exisiting large silicon companies including Intel and deep learning hardware startups are actively developing ASICs (Application Specific Integrated Circuit) for speeding up deep learning performance for training & inference. 
+
+Second approach is to deliever performance via software optimization, and nGraph Compiler stack is our solution to the second apporach. We created nGraph compiler stack not only to provide software acceleration for the upcoming deep learning ASICs but also to unlock massive performance boost for existing hardware targets such as CPUs and GPUs. Additionally, it aims to address the scalability issue with kernel libraries which is a popular solution to accelerate deep learning performance.
```
Comment (indie) [inline doc/sphinx/source/project/introduction.rst]: Typo on "deliver"
 
>  We created nGraph Compiler stack not only to provide software acceleration for upcoming deep learning ASICs but also to unlock massive performance boost for existing hardware targets such as CPUs and GPUs. 

Maybe a better way to phrase this is something like the nGraph approach is to enable people who want to write software/applications *today* (using CPUs and GPUs) that can and will easily benefit from custom ASICs or other datacenter deployments.  The software writer should not be worried  performance at all when they're just trying to get things to work the first time...   Datacenter admins can use nGraph to get more throughput out of their networks, regardless if users are running training or inference, with these customs ASICs.  Or, for totally local custom-built server racks, adding a custom ASIC like NNP provides more secure isolation from networks where sensitive data is being handled.

File: doc/sphinx/source/project/introduction.rst [hunk 3]
```diff
@@ -4,176 +4,86 @@
...
+
+Second approach is to deliever performance via software optimization, and nGraph Compiler stack is our solution to the second apporach. We created nGraph compiler stack not only to provide software acceleration for the upcoming deep learning ASICs but also to unlock massive performance boost for existing hardware targets such as CPUs and GPUs. Additionally, it aims to address the scalability issue with kernel libraries which is a popular solution to accelerate deep learning performance.  
 
 Motivations
 ===========
 
-Kernel libraries do not support graph-level optimizations
----------------------------------------------------------
+The current state of art software solution for speeding up deep learning computation is to integrate kernel libraries such as Intel MKL-DNN and NVidia's CuDNN into deep learning frameworks. These kernel libraries offer runtime performance boost on specific hardware targets through highly optimized kernels and operator level optimizations.
```
Comment (indie) [inline doc/sphinx/source/project/introduction.rst]: To achieve state-of-the-art performance with DL, the current solution is to integrate kernel libraries such as ...

File: doc/sphinx/source/project/introduction.rst [hunk 4]
```diff
@@ -4,176 +4,86 @@
...
-   :width: 555px
-   :alt: 
-
-   Figure A: Lack of graph-level optimization makes framework-to-kernel library
-   integration enormously inefficient. The computation graph above represents 
-   the computation: "A plus B times C".
-
-
-.. _figure-B:
+nGraph Compiler stack is designed to address the first two problems. nGraph applies graph level optimizations by taking computational graph from deep learning frameworks and reconstructing it with nGraph IR (Intermediate Representations). nGraph IR centralizes computational graphs from various frameworks and provides a unified way to connect backends for targetted hardware. To address the third problem, nGraph Compiler stack is integrated with PlaidML to generate code in LLVM, OpenCL, OpenGL, Cuda and Metal with low level optimizations automatically applied.
```
Comment (indie) [inline doc/sphinx/source/project/introduction.rst]: The nGraph Compiler stack addresses the first two issues.  It allows for both automatic and manual graph-level optimizations via the nGraph IR, which centralizes... 

typo on targeted

File: doc/sphinx/source/project/introduction.rst (hunk 1 above)
Comment (diyessi) [reply doc/sphinx/source/project/introduction.rst]: Agree; the sentences need some touching up.
Acceleration can come from faster execution as well as decreases in the time it takes to make new hardware and libraries available in frameworks.

File: doc/sphinx/source/project/introduction.rst (hunk 2 above)
Comment (diyessi) [reply doc/sphinx/source/project/introduction.rst]: Model once run everywhere, rather than adjust the model for each new device, is a goal.

File: doc/sphinx/source/project/introduction.rst (hunk 3 above)
Comment (diyessi) [reply doc/sphinx/source/project/introduction.rst]: Everyone want to be an editor :) `Deep learning frameworks currently derive performance from kernel libraries such as...`

File: doc/sphinx/source/project/introduction.rst (hunk 4 above)
Comment (diyessi) [reply doc/sphinx/source/project/introduction.rst]: Long term, we hope to address 3 with things like PlaidML.

File: doc/sphinx/source/project/introduction.rst [hunk 5]
```diff
@@ -4,176 +4,86 @@
...
+---------------------------------------------------------
 
-Integrating a framework on nGraph can be an attractive option for hardware 
-companies trying to design their own deep learning hardware or network architecture. 
-Framework integration is non-trivial amount of work, and nGraph automatically 
-does much of the heavy lifting. Furthermore, PlaidML can provide a wide range of 
-hardware coverage and optimization automatically. Any hardware that supports 
-LLVM, OpenCL, OpenGL, CUDA or Metal can be supported automatically with PlaidML 
-and nGraph.  
+As mentioned in the pervious section, kernel libraries need to be integrated with multiple deep learning frameworks, and this arduous task becomes even harder due to increased numbers of required kernels for achieving optimial performance. The number of required kernels is product of number of chip designs, data types, operations, and the cardinality of each parameter for each operation. In the past, the number of required kernels was limited, but as the AI research and industry rapidly develops, the final product of required kernels is increasing exponentially.
```
Comment (indie) [inline doc/sphinx/source/project/introduction.rst]: Delete "As mentioned in the pervious section".  

To transition in to this last part, maybe say something like "All of this adds up to arduous work as kernel libraries need to be integrated..."

File: doc/sphinx/source/project/introduction.rst [hunk 6]
```diff
@@ -4,176 +4,86 @@
...
 
-.. figure:: ../graphics/graph-compilers-at-a-glance.png
-   :width: 700px
-   :alt: Overview of various graph and tensor compilers.
+.. figure:: ../graphics/intro_kernel_explosion.png
+   :width: 555px
+   :alt: 
 
-   Figure G: Overview of various graph and tensor compilers.
+PlaidML was designed to address the expoential growth of kernel needs. It takes two inputs: operation defined by the user and machine description of the targetted hardware. It utilizes a Domain Specific Language (DSL) called Tile which allows developers to express how an operation should calculate tensors in a intutitive mathematical form. PlaidML takes user defined Tile code along with targed machine description such as threads, max memory input, etc to automatically apply low level optimizations. This automated optimization does not require kernel libraries to be written and lifts heavy burden for kernel developers. It also provides flexibility to support newer deep learning models in absence of hand optimized kernels for the new operations.
```
Comment (indie) [inline doc/sphinx/source/project/introduction.rst]: PlaidML addresses the kernel explosion problem in a manner that lifts a heavy burden off kernel developers.  It automatically lowers networks from nGraph into Tile, a :abbr:`Domain-Specific Language (DSL)` designed for deep learning that allows developers to express how an operation should calculate tensors in a intuitive mathematical form.  Integration of PlaidML with nGraph means extra flexibility to support newer deep learning models in absence of hand optimized kernels for the new operations.

**Comment by indie** [comment]
Since I updated this branch to the latest in master, it apparently triggered a bunch of stuff that wasn't merged before that is now merged; please feel free to ignore this request for review if you were erroneously added.  Feedback, is welcomed!

File: doc/sphinx/source/project/introduction.rst [hunk 7]
```diff
@@ -4,165 +4,191 @@
...
+datasets, data that must be encrypted, data-in-motion, and of course, in 
+finding the best compromises among speed, accuracy, and performance.  
+
+Two general approaches to advancing deep learning performance dominate the 
+industry today. The first is to design hardware dedicated exclusively to 
+handling compute for specialized kinds of :abbr:`Machine Learning (ML)` or 
+:abbr:`DL (Deep Learning)` operations; this approach essentially designs a 
+custom network infrastructure *around* specific problems AI is supposed to 
+solve. For example, many companies are actively developing specialized 
+:abbr:`ASICs (Application Specific Integrated Circuits)` to speed-up
```
Comment (rkimballn1) [inline doc/sphinx/source/project/introduction.rst]: In all other places were you have a abbr you have the words first with the abbreviation following inside parenthesis. Here you have the abbreviation first followed by the words being abbreviated.

File: doc/sphinx/source/project/introduction.rst [hunk 8]
```diff
@@ -4,165 +4,191 @@
...
 
-.. figure:: ../graphics/framework-to-kernel-lib.png
-   :width: 555px
-   :alt: 
+However, kernel libraries have three main problems: 
 
-   Figure A: Lack of graph-level optimization makes framework-to-kernel library
-   integration enormously inefficient. The computation graph above represents 
-   the computation: "A plus B times C".
+#. Kernel libraries do not support graph level optimizations.
```
Comment (rkimballn1) [inline doc/sphinx/source/project/introduction.rst]: hyphenate graph-level

File: doc/sphinx/source/project/introduction.rst [hunk 9]
```diff
@@ -4,165 +4,191 @@
...
+#. Framework integration of kernel libraries does not scale.
+#. There are too many kernels to write, and they require expert knowledge.
 
+The nGraph Compiler stack is designed to address the first two problems. nGraph 
+applies graph-level optimizations by taking the computational graph from a deep 
+learning framework like TensorFlow* and reconstructing it with the nGraph 
+:abbr:`Intermediate Representation (IR)`. The nGraph IR centralizes computational 
+graphs from various frameworks and provides a unified way to connect backends 
+for targeted hardware. From here, PlaidML or one of the nGraph custom 
+transformers can generate code various forms including LLVM, OpenCL, OpenGL,
```
Comment (rkimballn1) [inline doc/sphinx/source/project/introduction.rst]: I'm not sure what a "code various forms" is. Is it code-various forms? or something else.

File: doc/sphinx/source/project/introduction.rst [hunk 10]
```diff
@@ -4,165 +4,191 @@
...
-dynamically) into serialized graphs. Those graphs can then passed through a 
-compilation process (the *Graph optimization or compilation* step in 
-*Figure D* below), where various graph-level optimizations, like constant folding 
-or fusion can happen. These processes require unique vendor-provided libraries 
-to communicate with a driver (possibly through OpenCL\*, CUDA\*, or SYCL\*), to 
-compile and execute an implementation (kernel) for a specific 
-:abbr:`Instruction Set Architecture (ISA)`, or :term:`ISA`.
+The following computation is constructed to execute ``(A+B)*C``, but in the 
+context of nGraph, we can further optimize the graph to be represented as A*C. 
+From the first graph shown on the left, the operation on the constant B be
```
Comment (rkimballn1) [inline doc/sphinx/source/project/introduction.rst]: "constant B be can be" -> "constant B can be".

File: doc/sphinx/source/project/introduction.rst [hunk 11]
```diff
@@ -4,165 +4,191 @@
...
-compile and execute an implementation (kernel) for a specific 
-:abbr:`Instruction Set Architecture (ISA)`, or :term:`ISA`.
+The following computation is constructed to execute ``(A+B)*C``, but in the 
+context of nGraph, we can further optimize the graph to be represented as A*C. 
+From the first graph shown on the left, the operation on the constant B be 
+can be computed at the compile time (known as constant folding), and the 
+graph can be further simplified to the one on the right because the constant 
+has value of zero. Without such graph-level optimizations, a deep learning 
+framework with a kernel library will compute all operations, and the resulting 
+execution will be sub-optimal.
```
Comment (rkimballn1) [inline doc/sphinx/source/project/introduction.rst]: don't hyphenate suboptimal

File: doc/sphinx/source/project/introduction.rst [hunk 12]
```diff
@@ -4,165 +4,191 @@
...
+has value of zero. Without such graph-level optimizations, a deep learning 
+framework with a kernel library will compute all operations, and the resulting 
+execution will be sub-optimal. 
+
+
+Problem: Reduced scalability 
+----------------------------
+
+Integrating kernel libraries with frameworks is increasingly becoming 
+non-trivial due to the growing number of new deep learning accelerators.
```
Comment (rkimballn1) [inline doc/sphinx/source/project/introduction.rst]: don't hypenate nontrivial

File: doc/sphinx/source/project/introduction.rst [hunk 13]
```diff
@@ -4,165 +4,191 @@
...
+learning curve in taking their concepts from design to production to scale. The 
+fundamental efficiencies behind Moore's Law are not dead; rather than fitting 
+`more transistors on denser and denser circuits`_, we're enabling advances 
+in compute with more transformers on denser and more data-heavy 
+:abbr:`Deep Learning Networks (DNNs)`, and making it easier to apply advances in 
+:abbr:`Machine Learning (ML)` to different of industries and problems. 
 
-Other notable efforts
-----------------------
+For developers with a network already in-place, using the nGraph Compiler stack
```
Comment (rkimballn1) [inline doc/sphinx/source/project/introduction.rst]: don't hypenate "in place"

File: doc/sphinx/source/project/introduction.rst [hunk 14]
```diff
@@ -4,165 +4,191 @@
...
 
-A few other notable efforts in compiler projects include: 
+This documentation provides technical details of nGraph's core functionality, framework 
+and backend integrations. Creating a compiler stack like nGraph and PlaidML 
+requires expert knowledge, and we hope nGraph and PlaidML will lift burden for 
 
-* **TVM** https://github.com/dmlc/tvm
-* **XLA** https://developers.googleblog.com/2017/03/xla-tensorflow-compiled.html
-* **Glow** https://arxiv.org/pdf/1805.00907.pdf 
+1. Framework owners looking to support new hardware and custom chips.
```
Comment (rkimballn1) [inline doc/sphinx/source/project/introduction.rst]: Can you use '#' instead of numbers like you did on line 65?

File: doc/sphinx/source/sitemap.rst [hunk 15]
```diff
@@ -3,6 +3,6 @@
 .. _sitemap.rst:
  
 .. toctree::
-   :includehidden:
- 
+   :includehidden:
```
Comment (rkimballn1) [inline doc/sphinx/source/sitemap.rst]: Might want to just revert this change since it does not look like anything really changed.

File: doc/sphinx/source/project/introduction.rst (hunk 7 above)
Comment (indie) [reply doc/sphinx/source/project/introduction.rst]: So this is a second instance of the same term, which means that we want the abbreviation to render to the reader, not the spelled-out acronym.  When they hover, it will show the full acronym, which is what we want.

File: doc/sphinx/source/project/introduction.rst (hunk 8 above)
Comment (indie) [reply doc/sphinx/source/project/introduction.rst]: :+1:

File: doc/sphinx/source/sitemap.rst (hunk 15 above)
Comment (indie) [reply doc/sphinx/source/sitemap.rst]: Yeah, weird merge diff from fixing this yesterday for the OTC folks.

**Comment by aprocter** [comment]
Unfortunately I don't think I'm going to have time to go through this.
